src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -934,16 +934,19 @@ internal ValueTask<FlushResult> WriteAsync(ReadOnlyMemory<byte> source, Cancella
             // state as writing
             AllocateWriteHeadIfNeeded(0);
 
-            if (source.Length <= _writingHeadMemory.Length)
+            lock (_sync)
             {
-                source.CopyTo(_writingHeadMemory);
+                if (source.Length <= _writingHeadMemory.Length)
+                {
+                    source.CopyTo(_writingHeadMemory);
 
-                AdvanceCore(source.Length);
-            }
-            else
-            {
-                // This is the multi segment copy
-                WriteMultiSegment(source.Span);
+                    AdvanceCore(source.Length);
+                }
+                else
+                {
+                    // This is the multi segment copy
+                    WriteMultiSegment(source.Span);
+                }
```

Comment:
Is there any concern of acquiring and releasing the lock immediately between here and FlushAsync? Can a call interleave in here that would affect the write call? I doubt it, but it's the only cause for concern I see.

Reply:
Should be fine. Efficiency.